apollo-mockserver/src/commonMain/kotlin/com/apollographql/apollo3/mockserver/MockServerCommon.kt

```diff
@@ -51,11 +50,16 @@ class MockResponse(
   constructor(
       body: String,
       statusCode: Int = 200,
-      headers: Map<String, String> = emptyMap()
+      headers: Map<String, String> = emptyMap(),
   ) : this(statusCode, body.encodeUtf8(), headers)
 }
 
-internal fun readRequest(source: BufferedSource): MockRecordedRequest? {
+interface MockServerHandler {
+  fun handle(request: MockRequest): MockResponse
+  fun copy(): MockServerHandler
```

Comment:
Not a huge fan of the `copy()` because:

1. it's a K/N detail that JVM implementers shouldn't need to worry about.
2. in some scenarios, it makes it possible to "miss" some enqueued response. For an example if enqueuing a response while another one is being handled, the queue weil be overwritten.

I think we can make a `QueueMockDispatcher` that survives freezing by using a `NSMutableArray`  under the hood, like in `Socket.kt`. Let me know if you want to try, if not, I'll give it a shot.

Reply:
Yeah I agree it's less than ideal!

Ok I'll try making `QueueMockDispatcher` `expect` with a specific implementation for Apple. 

There's still an expectation that custom implementations must be "freeze safe" - but a comment on the interface may be enough?  Something like _"Note that on Apple platform, instances will be frozen, and thus need to be immutable (this is not required for other platforms)"_?

Reply:
I think this is a reasonable expectation. No need to mention `freeze` explicitely, especially if we're planning to remove them in the mid-term. I'd focus at the "multi-threaded" portion. Then every platform can do what they want (freeze, lock, remove all side effects, etc...)

```
  /**
   * handles the given [MockRequest]
   * 
   * This method is called from one or several background thread and should be thread-safe.  
   */
  fun handle(request: MockRequest): MockResponse
```

Reply:
mmmm and re-reading my comment, I just understand what you meant 🤦 ... Maybe we can remove the `freeze()` from the apollo code? Leave it to the implementation instead? 

`QueueMockServerHandler` would just "freeze" itself from its constructor. Someone using the new MM (or just having a pure handler) could skip this

Reply:
The thing is `MockServerHandler` is not frozen explicitly but "transitively" as it's a property of `Socket` which needs to be frozen. 

I actually agree with your comment on not mentioning `freeze` explicitly and instead talking about threading, I think it makes sense.

Making an Apple specific `QueueMockDispatcher` using `NSMutableArray` did work! 💪  (used `expect`/`actual` in the end) - coming in the next commit.

Reply:
> The thing is MockServerHandler is not frozen explicitly but "transitively" as it's a property of Socket which needs to be frozen.

Right. I see... We _could_ expose a `Factory` to move the creation to a later point in time. But TBH I don't think it's worth the extra complexity. Let's start with freezing everything and if we need to we can always introduce factories later on?